### PR TITLE
NEW Use x.y-release branch instead of detached head

### PR DESCRIPTION
**Pull request by GuySartorelli**
## New release branch
When performing a new minor release, cow will now create your new minor branch _and_ the release branch. It will then add commits to that branch instead of creating them in a detached head for tagging.

e.g. if releasing 4.12.0-beta1 cow will:
1. create a `4.12` branch
2. create a `4.12-release` branch
3. create a commit on the new `4.12-release` branch updating composer dependencies from `4.x-dev` to `~4.12.0@beta`
4. push the `4.12-release` branch up to the remote
5. tag the `4.12-beta1` release
6. push the new tag
7. finally, push the `4.12` branch up to the remote

Note that actual numbers here may differ depending on the actual version of the recipe or module being released (e.g. 4.12 would be 1.12 for recipe-blog).

## New merge up command
Running `cow release:mergeup` will merge the release branch up to the minor branch, and the minor branch up to the major branch, updating the constraints as it goes. You can optionally pass `--delete-release-branch` and the `x.y-release` branch will be deleted on the remote afterward.

e.g. after releasing 4.12.0 _stable_, running `cow release:mergeup --delete-release-branch` will:
1. Make a commit updating composer dependencies from `~4.12.0@stable` to `4.12.x-dev`
2. Merge `4.12-release` to `4.12` and push to github
3. Make a commit updating composer dependencies from `4.12.x-dev` to `4.x-dev`
4. Merge `4.12` to `4`
5. Delete the `4.12-release` branch from github

Note that actual numbers here may differ depending on the actual version of the recipe or module being released (e.g. 4.12 would be 1.12 for recipe-blog).

## Parent issue
- https://github.com/silverstripeltd/product-issues/issues/565

### CODE REVIEW

File: src/Steps/Release/PublishRelease.php
Comment (emteknetnz) [inline src/Steps/Release/PublishRelease.php:139]:
```suggestion
            $parentLibrary->setComposerData($composerData, true, 'DEP Update release dependencies');
```
I'd the stabilise commit should have a 'DEP' and the dev commit should have a MNT

File: src/Steps/Release/MergeUpRelease.php
Comment (emteknetnz) [inline src/Steps/Release/MergeUpRelease.php:123]:
```suggestion
            // a Gitonomy\Git\Exception\ProcessException will be thrown and cow will exit
```
So, lets say this happens ... what now?

**Review by emteknetnz — CHANGES_REQUESTED** [review]
No unit tests?

File: src/Steps/Release/PublishRelease.php
Comment (GuySartorelli) [reply src/Steps/Release/PublishRelease.php:139]: The stabilise commits _already_ has `MNT` - this PR isn't going to change those. If you think those should change, you can create a separate card to handle that unrelated change.

File: src/Steps/Release/MergeUpRelease.php
Comment (GuySartorelli) [reply src/Steps/Release/MergeUpRelease.php:123]: The release manager should resolve the conflict, then checkout the release branch again, and re-run the command.

Cow will still go through and process the branches it's already done - but it will find there is nothing to merge, and the composer changes will match what's already there, so no new commits will be added, so it won't cause any changes to be made on the modules which were already processed.
It will then get to the one where the release manager resolved the conflicts and continue from there.

**Comment by GuySartorelli** [comment]
> No unit tests?

Most of cow doesn't have unit tests.... when do we write tests for this and when do we not?
I'm not even sure _how_ to write a test for this new behaviour. Certainly there's no precedent for testing actual git steps with the existing unit tests.

File: src/Steps/Release/MergeUpRelease.php
Comment (emteknetnz) [inline src/Steps/Release/MergeUpRelease.php:93]: I'm not sure how much to worry about automatically merging up JS at this point. This could be a real blocker to doing this via cow.

What we should be doing here, is if there were JS src changes we'd be checking out silverstripe/admin at the corresponding version, running yarn install there, and running yarn build in the current module

We do something similar in CI when running js tests - https://github.com/silverstripe/gha-run-tests/blob/1.0/action.yml#L184

File: src/Steps/Release/MergeUpRelease.php
Comment (emteknetnz) [inline src/Steps/Release/MergeUpRelease.php:164]: What happens when we have dual support constraints such as graphql 3 and 4?  Yes it's a "temporary" hack, though we may also put in similar hacks in the future, or simply we forced in to dual support, and fail to update this here.  I think it's safer to use something like
```php
$contraint = str_replace('@stable', '', $currentConstraint);
$contraint = preg_replace('#([0-9]+\.[0-9]+)\.[0-9]+#', '$1', $constraint) . '.x-dev';
```

File: src/Steps/Release/MergeUpRelease.php
Comment (emteknetnz) [inline src/Steps/Release/MergeUpRelease.php:55]: I'd be inclined to exit at this point. The user may have done something like manually switch a branch in their release dir and forgot to change it back.

If we simply skip here, then it's possible that we'll miss that an entire recipe wasn't merged up properly

I guess it does get complex if we've previously exited because there was an issue with any merge-ups where we'll exit, so, I do understand the motivation to skip here

I'm thinking perhaps we should move the -release branch deletion to its own step and do that at the very end.  That way we can reliably do detection at this point if there is an existing -release branch that we should possibly switch to?

File: src/Steps/Release/MergeUpRelease.php
Comment (emteknetnz) [inline src/Steps/Release/MergeUpRelease.php:69]: As noted above, I think this should be its own step

**Review by emteknetnz — CHANGES_REQUESTED** [review]

File: src/Steps/Release/PublishRelease.php
Comment (GuySartorelli) [reply src/Steps/Release/PublishRelease.php:89]: This was from before I added a check against this earlier in the process. Will remove.

File: src/Steps/Release/MergeUpRelease.php
Comment (GuySartorelli) [reply src/Steps/Release/MergeUpRelease.php:25]: This is unnecessary. `delete-release-branch` is a `InputOption::VALUE_NONE` type option - which means it can only be true or false. I think the null coallescing operator there is redundant as well.

File: src/Model/Modules/Library.php
Comment (GuySartorelli) [reply src/Model/Modules/Library.php]: Done

File: src/Model/Modules/Library.php
Comment (GuySartorelli) [reply src/Model/Modules/Library.php]: Done

File: src/Steps/Release/MergeUpRelease.php
Comment (GuySartorelli) [reply src/Steps/Release/MergeUpRelease.php]: Done

File: src/Steps/Release/MergeUpRelease.php
Comment (GuySartorelli) [reply src/Steps/Release/MergeUpRelease.php]: Done

File: src/Commands/Release/MergeUp.php
Comment (GuySartorelli) [reply src/Commands/Release/MergeUp.php:45]: Done

File: src/Steps/Release/PublishRelease.php
Comment (GuySartorelli) [reply src/Steps/Release/PublishRelease.php]: Done

File: src/Steps/Release/MergeUpRelease.php
Comment (GuySartorelli) [reply src/Steps/Release/MergeUpRelease.php:163]: > Yes it's a "temporary" hack

I don't want to bake solutions for temporary hacks into this. IF we're going to provide functionality to support dual constraints it needs to be across the board and the graphql hack needs to be done properly and merged in.
Otherwise I don't see the point in making some special handling here that doesn't actually matter because cow would never get to this point because the plan can't support dual constraints and the stabalise constraint functionality also can't handle dual constraints.

File: src/Steps/Release/MergeUpRelease.php
Comment (GuySartorelli) [reply src/Steps/Release/MergeUpRelease.php]: What would a release branch look like for a patch release? Would we even have one? If so, would it be e.g. 4.12-release for a 4.12.6 patch release? If so this logic works fine for that scenario, I think?

For a major release, would that not just effectively be the same as a minor as far as branches (e.g. create 5.0-release branch, merge up to 5.0 and that up to 5)?

File: src/Steps/Release/MergeUpRelease.php
Comment (GuySartorelli) [reply src/Steps/Release/MergeUpRelease.php:159]: I would expect that for a major release at least we'd still be doing the same constraint changes that we do for minors, and the branching strategy would be the same as mentioned in https://github.com/silverstripe/cow/pull/226#discussion_r1022195852

I don't know enough about how cow would be used for a patch release to know what changes you want me to make for that scenario.

File: src/Steps/Release/PublishRelease.php
Comment (GuySartorelli) [reply src/Steps/Release/PublishRelease.php]: The release manager may have been doing something that required them to swap branches, and forgot to swap back. Probably it would be better to just bail in that case though, and tell the user to resolve the problem.
I'll do that instead.

File: src/Steps/Release/MergeUpRelease.php
Comment (GuySartorelli) [reply src/Steps/Release/MergeUpRelease.php:51]: I'd be much happier exiting here than having deletion as its own separate command (I assume that's what you mean by "step" here?), so that's the change I'll make for now.
I can split off deletion as its own command if you really want, or if you did actually mean a separate step in the same command, then I'll need more information about what you meant and how you expect that flow to go.

> I guess it does get complex if we've previously exited because there was an issue with any merge-ups where we'll exit, so, I do understand the motivation to skip here

I don't think it does - after fixing issues with merge-ups, the release manager should checkout the release branch again before running the command again. So having this check is a good protection against forgetting to do that.

>  That way we can reliably do detection at this point if there is an existing -release branch that we should possibly switch to?

We could do that here anyway - but we will already be on the release branch, because that was the last thing to happen as part of the merging. I can put in another `if (!$releasePlanNode->isOnReleaseBranch())` check before deletion though if you like.

File: src/Steps/Release/MergeUpRelease.php
Comment (GuySartorelli) [reply src/Steps/Release/MergeUpRelease.php:120]: I would be shocked if someone thought adding a global exception handler to cow was a good idea, it uses exceptions to bail in many places, so that would already cause problems.

The exception already includes the git error message, which in the case of a commit will look like this (obviously a conflict in composer.json is unlikely but this gets the idea across):
> Auto-merging composer.json
> CONFLICT (content): Merge conflict in composer.json
> Automatic merge failed; fix conflicts and then commit the result.

I don't think we need to be any more explicit about what commands to use than this - anyone doing a release already knows how to resolve merge conflicts. Besides which it's concievable that the merge could fail for other reasons other than a merge conflict (though I admittedly can't think of such a scenario right now) - in which case the exception will say what went wrong, but adding messaging saying to resolve a merge conflict that doesn't exist would be confusing.

I'll add more explicit messaging if you want, and I'll catch the exception and call `die()` instead if you want, but I don't think either of those are necessary.

File: src/Steps/Release/MergeUpRelease.php
Comment (emteknetnz) [reply src/Steps/Release/MergeUpRelease.php]: I honestly don't know, I've only every used cow end to end for minor releases

I suppose for patch releases it would be from the same 4.12 branch and we'd still need to update and revert the stable deps and then do the double merge-up

For majors ... I really don't know if it uses 4 or 4.12 as the base branch before going to 5 / 5.0

File: src/Steps/Release/MergeUpRelease.php
Comment (GuySartorelli) [reply src/Steps/Release/MergeUpRelease.php]: If there are JS changes to merge up, cow will bail when the dist bundle merge conflict innevitably occurs, and the release manager can then do those steps manually.
I think this is probably safer than automating js merge-ups - consider for example if there's a merge conflict in some file other than the bundle.js in the same merge - we'd have to detect that separately.

File: src/Steps/Release/MergeUpRelease.php
Comment (emteknetnz) [reply src/Steps/Release/MergeUpRelease.php:163]: We'll need to retain the "hack", CMS 4 will always retain dual support for graphql, and we need to be retain the ability to release 4.14 in the future if something special comes up.

I guess raise a new PR for whatever hacks were used in the 4.12 beta release?

File: src/Steps/Release/MergeUpRelease.php
Comment (emteknetnz) [reply src/Steps/Release/MergeUpRelease.php]: > as its own separate command

yes that's what I meant.  To me it seems much safer to do -release branch deletion at the very end of the cow process, because it's technically an optional step and you could actually just leave those old branches lying around without much real world impact.  Retaining the -release branches during the release is completely safe, and they can be used within cow to detect if there was a release branch created, and if they differ from the dev branch.  This can be used as protection against things like "I manually changed to the dev branch from the release branch and forgot to change back"

File: src/Steps/Release/MergeUpRelease.php
Comment (emteknetnz) [reply src/Steps/Release/MergeUpRelease.php:129]: I think we still need the explict messaging with instructions.  Cow is a bit of black box and the person managing the release may not fully grasp the 3x branches and stable vs dev dependencies in recipes (I know it seems obvious now, but it took me multiple releases to really get my head around this, though to be fair most of the confusion was because of --detach-tagged-base)

For instance, simply saying `git merge --no-ff 1.12-release` is going to be valuable in this context so the user knows they're merge from the correct branch.

File: src/Steps/Release/MergeUpRelease.php
Comment (emteknetnz) [reply src/Steps/Release/MergeUpRelease.php:168]: I don't know either, you'll need to look into this further.  We're going to use cow to release `5.0.0` so this needs to work correctly.

File: src/Steps/Release/MergeUpRelease.php
Comment (emteknetnz) [reply src/Steps/Release/MergeUpRelease.php]: Yeah I guess this is fine. If we mess up the js on merge-up, it'll only affect dev branches and show in CI.  The actual tagged release won't ship with merge conflicts

File: src/Steps/Release/MergeUpRelease.php
Comment (GuySartorelli) [reply src/Steps/Release/MergeUpRelease.php]: Okay. So we can agree that for patch releases it's the same flow as for the minor release (e.g. releasing a 4.12.999 patch needs the same stuff to happen as releasing a 4.12.0 minor as far as this PR is concerned).

I think that's true for majors as well. Releasing 5.0.0 is the same as releasing 5.1.0.
- a new `5.0` branch is created from `5`
- a new `5.0-release` branch is created from `5.0`
- constraints are updated in the `5.0-release` branch to be `~5.0.0@stable` (after having been `@beta` and `@rc`)
- all the normal release stuff happens. If there are any changes needed in this part, it's beyond the concerns of this PR
- after the release, merge-up starts
- merge up from `5.0-release` to `5.0` and destabalise constraints to `5.0.x-dev`
- merge up from `5.0` to `5` and destabalise constraints to `5.x-dev`
- delete release branch

I don't think this PR needs to change _at all_ for all of that to work.

File: src/Steps/Release/MergeUpRelease.php
Comment (GuySartorelli) [reply src/Steps/Release/MergeUpRelease.php:168]: See the comment linked above. I don't think any changes are required to this PR for it to work for a patch or a major release.

File: src/Steps/Release/MergeUpRelease.php
Comment (GuySartorelli) [reply src/Steps/Release/MergeUpRelease.php:172]: I've reopened https://github.com/silverstripe/cow/issues/207 to handle that. It's beyond the concerns of this PR.

File: src/Steps/Release/MergeUpRelease.php
Comment (GuySartorelli) [reply src/Steps/Release/MergeUpRelease.php:129]: Cow will already log out "Merging library <info>{$name}</info> from <info>{$mergeFrom}</info> to <info>{$mergeInto}</info>"
This will result in something like:

> Merging library `silverstripe/framework` from `4.12-release` to `4.12`

Then the exception message from git will say (along with some debug info) the message in my previous comment above. So the context for what repo is being dealt with, what went wrong, what branch is being merged into what other branch, that's all there already.
When checking out the directory for that repo, you'll be in the conflict state already so running `git merge --no-ff 1.12-release` is unnecessary.

I'm adding messaging for what commands to run after resolving the conflict as we discussed. It's concievable that there will be some scenario where this is _not_ the correct set of commands to run, but since I can't think of a specific scenario I'll just let it be.

File: src/Steps/Release/MergeUpRelease.php
Comment (GuySartorelli) [reply src/Steps/Release/MergeUpRelease.php]: Done

File: src/Steps/Release/MergeUpRelease.php
Comment (emteknetnz) [reply src/Steps/Release/MergeUpRelease.php]: Right, sorry I think I confused myself with "releasing majors".  In hindsight it's very obviously going to use the existing `5` branch, just as it uses the existing `4` branch for minors.  We've created these branches ourselves.  There's possibly some existing functionality within cow where it would create the new major branches, but we won't use this anytime soon.

**Review by emteknetnz — CHANGES_REQUESTED** [review]

File: src/Steps/Release/MergeUpRelease.php
Comment (GuySartorelli) [reply src/Steps/Release/MergeUpRelease.php:135]: What's wrong with using tabs to indent?

File: src/Steps/Release/MergeUpRelease.php
Comment (GuySartorelli) [reply src/Steps/Release/MergeUpRelease.php:109]: An `InvalidArgumentException` is thrown ("Branch {$branch} is not a local or remote branch"), and progress is halted.

File: src/Steps/Release/MergeUpRelease.php
Comment (GuySartorelli) [reply src/Steps/Release/MergeUpRelease.php:117]: `$mergeFrom` is the release branch (which we _know_ exists because we've already checked we're on that branch) the first time and then the minor branch (which we _also_ know exists because we just checked it out and merged into it) the second time.

So you won't ever actually be in the situation that `$mergeFrom` doesn't exist by the time you hit this - but if you did, a `ProcessException` would be thrown from the git library with a message like "fatal: ambiguous argument 'some-branch': unknown revision or path not in the working tree."

File: src/Steps/Release/MergeUpRelease.php
Comment (GuySartorelli) [reply src/Steps/Release/MergeUpRelease.php:83]: I can't see any value in doing that because a) we'll never be in that situation (how would we have a `4.12-release` branch without a corresponding `4.12` and `4` branch?) and b) even if we somehow _did_ get in that situation, cow will find that when it tries to checkout the branches, and halt at that stage.

I'll add the checks if you insist, but I don't think it's necessary.

File: src/Commands/Release/DeleteReleaseBranch.php
Comment (GuySartorelli) [reply src/Commands/Release/DeleteReleaseBranch.php:39]: Done

File: src/Application.php
Comment (GuySartorelli) [reply src/Application.php:90]: Oops yes

File: src/Steps/Release/MergeUpRelease.php
Comment (emteknetnz) [reply src/Steps/Release/MergeUpRelease.php:135]: > What's wrong with using tabs to indent?

ლ(ಠ益ಠლ)

Comment (emteknetnz) [reply src/Steps/Release/MergeUpRelease.php:135]: Tabs are beautiful :-)

File: src/Steps/Release/MergeUpRelease.php
Comment (emteknetnz) [reply src/Steps/Release/MergeUpRelease.php:83]: My main concern would be if $releaseBranch ($library->getBranch()) wasn't actually a release branch because it was manually checked out by someone for whatever reason

Comment (GuySartorelli) [reply src/Steps/Release/MergeUpRelease.php:83]: That's already protected against in `mergeRecursive()` - but I'll add explicit checks so there's no ambiguity.

Comment (emteknetnz) [reply src/Steps/Release/MergeUpRelease.php:83]: Talked about offline, there should be no way that the branches are not what's expected due to validation that we're on a release branch on the previous method that calls this method

**Review by emteknetnz — APPROVED** [review]
Awesome work!

**Pull request merged**
